pandas/core/indexes/base.py

```diff
@@ -4550,11 +4550,16 @@ def putmask(self, mask, value):
             return self.astype(dtype).putmask(mask, value)
 
         values = self._values.copy()
-        if isinstance(converted, np.timedelta64) and self.dtype == object:
+        dtype, _ = infer_dtype_from(converted, pandas_dtype=True)
+        if dtype.kind in ["m", "M"]:
             # https://github.com/numpy/numpy/issues/12550
             #  timedelta64 will incorrectly cast to int
-            converted = [converted] * mask.sum()
-            values[mask] = converted
+            if not is_list_like(converted):
```

Comment:
this pattern is pretty common, maybe a helper function?

Reply:
yah one more PR coming up that uses this pattern yet again; i think its going to end up in array_algos.putmask,